src/api/c++/z3++.h

```diff
@@ -448,14 +485,23 @@ namespace z3 {
     public:
         param_descrs(context& c, Z3_param_descrs d): object(c), m_descrs(d) { Z3_param_descrs_inc_ref(c, d); }
         param_descrs(param_descrs const& o): object(o.ctx()), m_descrs(o.m_descrs) { Z3_param_descrs_inc_ref(ctx(), m_descrs); }
+        param_descrs(param_descrs && o) noexcept : object{std::forward<object>(o)}, m_descrs{o.m_descrs} {
```

Comment:
std::move is so much more concise. forward is for templates

Reply:
Normally, but in this case we are not moving the entire instance. Applying move semantics to the entire thing leads to the implication that the entire instance is to-be consumed. But here we want only the `object` part of the instance to be consumed; in other words the cast is a promise that only the desired components of the instance will be consumed rather than all of it which move semantics traditionally imply.

Reply:
std::move doesn't move. There's no implication.
std::forward makes the code harder to maintain. If you change the base class you need to update all forwards.

Reply:
I might have been unclear: `std::move` casts to a r-value reference; the implication of passing a move reference is that the function accepting it will consume it. As for the base class, if it changes then this line must be updated anyway since `object{` would no longer be valid.

Reply:
std::move doesn't move. There are no implications.

Reply:
It might not explicitly move, but it has strong implications. Consider `foo(std::move(bar))`. This should imply that `foo` consumes `bar`. If not, `foo` should take `bar` in by reference, pointer, or value. Except with forwarding const references in templates, where you don't use `std::move` anyway, r-value references' primary use is moving an item. `std::move` doesn't move, but it is designed to indicate something can be move; it is an explicit cast that says 'this object may be moved', otherwise it wouldn't be called `std::move`. In most contexts, `foo(std::move(x))` implies that `foo` consumes `x`, because otherwise you shouldn't invoke `std::move`. Even here in the constructor that is what we are doing, `object`'s constructor consumes the `object` portion of the instance we are passing it. Substituting `std::move` is an example, if we have `object{std::move(s)}` object is clearly consuming `s`, that is why we use `std::move`. Thus things wrapped in `std::move`, after passed to a function, should be considered 'already moved / altered'. The difference here is that with `std::forward<object>(s)` we are not saying 'consume s', but rather 'consume the object part of s'. In other words, 'the non-object part of s will not be touched by the function and we can safely use the rest of `s` after this function call'.